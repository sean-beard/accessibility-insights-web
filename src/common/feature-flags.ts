// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { forEach } from 'lodash';

import { FeatureFlagStoreData } from './types/store-data/feature-flag-store-data';

export class FeatureFlags {
    public static readonly logTelemetryToConsole = 'logTelemetryToConsole';
    public static readonly showAllAssessments = 'showAllAssessments';
    public static readonly shadowDialog = 'shadowDialog';
    public static readonly showAllFeatureFlags = 'showAllFeatureFlags';
    public static readonly scoping = 'scoping';
    public static readonly showInstanceVisibility = 'showInstanceVisibility';
    public static readonly manualInstanceDetails = 'manualInstanceDetails';
    public static readonly debugTools = 'debugTools';

    public static readonly exportReport = 'exportReport';
}

export interface FeatureFlagDetail {
    id: string;
    defaultValue: boolean;
    displayableName: string;
    displayableDescription: string;
    isPreviewFeature: boolean;
    forceDefault: boolean;
}

export function getAllFeatureFlagDetails(): FeatureFlagDetail[] {
    return [
        {
            id: FeatureFlags.shadowDialog,
            defaultValue: false,
            displayableName: 'Improved dialog styling',
            displayableDescription:
                'Avoids styling problems in failure dialogs by rendering them in shadow DOM. ' +
                "(You'll need to refresh the target page to see the new dialog styling.)",
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.showAllAssessments,
            defaultValue: false,
            displayableName: 'Show all assessments',
            displayableDescription: 'Show all assessments, even the in-development ones',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.logTelemetryToConsole,
            defaultValue: false,
            displayableName: 'Log telemetry to console',
            displayableDescription:
                'Write telemetry payload information to the developer tools console.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.showAllFeatureFlags,
            defaultValue: false,
            displayableName: 'Show all feature flags',
            displayableDescription: 'Show all feature flags in the Preview Features panel.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.scoping,
            defaultValue: false,
            displayableName: 'Scoping experience',
            displayableDescription:
                'Enable scoping to limit scanning to selected portions of the webpage.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.showInstanceVisibility,
            defaultValue: false,
            displayableName: 'Show instance visibility in assessment',
            displayableDescription:
                'Shows visibility of instances in assessment requirement lists. May impact performance. ' +
                "(You'll need to go to a different requirement and come back for it to take effect.)",
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.manualInstanceDetails,
            defaultValue: false,
            displayableName: 'Enable manual instance details',
            displayableDescription:
                'Allow addition of path (CSS selector) which automatically ' +
                'populates the corresponding code snippet when adding manual failure instance.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.debugTools,
            defaultValue: false,
            displayableName: 'Enable debug tools',
            displayableDescription:
                'Click on the new icon close to the gear to open the debug tools',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.exportReport,
            defaultValue: false,
            displayableName: 'More export options',
            displayableDescription: 'Enables exporting reports to external services',
            isPreviewFeature: true,
            forceDefault: false,
        },
    ];
}

export function getDefaultFeatureFlagValues(): FeatureFlagStoreData {
    const details: FeatureFlagDetail[] = getAllFeatureFlagDetails();
    const values: FeatureFlagStoreData = {};
    forEach(details, detail => {
        values[detail.id] = detail.defaultValue;
    });
    return values;
}

export function getForceDefaultFlags(): FeatureFlags[] {
    const details: FeatureFlagDetail[] = getAllFeatureFlagDetails();
    const forceDefaultFlags: FeatureFlags[] = [];
    forEach(details, detail => {
        if (detail.forceDefault) {
            forceDefaultFlags.push(detail.id);
        }
    });
    return forceDefaultFlags;
}
