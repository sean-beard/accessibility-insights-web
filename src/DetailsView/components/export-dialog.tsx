// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FlaggedComponent } from 'common/components/flagged-component';
import { FeatureFlags } from 'common/feature-flags';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { Dialog, DialogFooter, DialogType, PrimaryButton, TextField } from 'office-ui-fabric-react';
import * as React from 'react';
import { ReportExportServiceProvider } from 'report-export/report-export-service-provider';
import { ExportFormat } from 'report-export/types/report-export-service';
import { ExportResultType } from '../../common/extension-telemetry-events';
import { FileURLProvider } from '../../common/file-url-provider';
import { NamedFC } from '../../common/react/named-fc';
import { DetailsViewActionMessageCreator } from '../actions/details-view-action-message-creator';

export interface ExportDialogProps {
    deps: ExportDialogDeps;
    isOpen: boolean;
    fileName: string;
    description: string;
    html: string;
    onClose: () => void;
    onDescriptionChange: (value: string) => void;
    exportResultsType: ExportResultType;
    onExportClick: () => void;
    featureFlagStoreData: FeatureFlagStoreData;
}

export interface ExportDialogDeps {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
    fileURLProvider: FileURLProvider;
    reportExportServiceProvider: ReportExportServiceProvider;
}

export const ExportDialog = NamedFC<ExportDialogProps>('ExportDialog', props => {
    const [format, setFormat] = React.useState<ExportFormat | null>(null);

    const onDismiss = (): void => {
        props.onClose();
    };

    const onExportLinkClick = (
        event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>,
        exportFormat: ExportFormat,
    ): void => {
        const { detailsViewActionMessageCreator } = props.deps;
        props.onDescriptionChange(props.description);
        detailsViewActionMessageCreator.exportResultsClicked(
            props.exportResultsType,
            props.html,
            event,
        );
        setFormat(exportFormat);
        props.onExportClick();
        props.onClose();
    };

    const onDescriptionChange = (event, value: string): void => {
        props.onDescriptionChange(value);
    };

    const fileURL = props.deps.fileURLProvider.provideURL([props.html], 'text/html');
    const exportService = props.deps.reportExportServiceProvider.forKey(format);
    const ExportForm = exportService ? exportService.exportForm : null;

    const getSingleExportToHtmlButton = () => {
        return (
            <PrimaryButton
                onClick={event =>
                    onExportLinkClick(event as React.MouseEvent<HTMLAnchorElement>, 'download')
                }
                download={props.fileName}
                href={fileURL}
            >
                Export
            </PrimaryButton>
        );
    };

    const getMultiOptionExportButton = () => {
        return (
            <>
                <PrimaryButton
                    text="Export"
                    split
                    splitButtonAriaLabel="Export HTML to any of these format options"
                    aria-roledescription="split button"
                    menuProps={{
                        items: props.deps.reportExportServiceProvider.all().map(service => ({
                            key: service.key,
                            text: service.displayName,
                            onClick: (e: React.MouseEvent<HTMLButtonElement>) => {
                                onExportLinkClick(e, service.key);
                            },
                        })),
                    }}
                    onClick={(e: React.MouseEvent<HTMLAnchorElement>) => {
                        onExportLinkClick(e, 'download');
                    }}
                    download={props.fileName}
                    href={fileURL}
                />
                {ExportForm && (
                    <ExportForm
                        fileName={props.fileName}
                        description={props.description}
                        html={props.html}
                        onSubmit={() => {
                            setFormat(null);
                        }}
                    />
                )}
            </>
        );
    };

    return (
        <Dialog
            hidden={!props.isOpen}
            onDismiss={onDismiss}
            dialogContentProps={{
                type: DialogType.normal,
                title: 'Provide result description',
                subText: 'Optional: please describe the result (it will be saved in the report).',
            }}
            modalProps={{
                isBlocking: false,
                containerClassName: 'insights-dialog-main-override',
            }}
        >
            <TextField
                multiline
                autoFocus
                rows={8}
                resizable={false}
                onChange={onDescriptionChange}
                value={props.description}
                ariaLabel="Provide result description"
            />
            <DialogFooter>
                <FlaggedComponent
                    featureFlag={FeatureFlags.exportReport}
                    featureFlagStoreData={props.featureFlagStoreData}
                    enableJSXElement={getMultiOptionExportButton()}
                    disableJSXElement={getSingleExportToHtmlButton()}
                />
            </DialogFooter>
        </Dialog>
    );
});
