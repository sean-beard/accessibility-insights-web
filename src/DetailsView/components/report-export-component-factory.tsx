// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { CommandBarProps } from 'DetailsView/components/details-view-command-bar';
import {
    ReportExportComponent,
    ReportExportComponentProps,
} from 'DetailsView/components/report-export-component';
import * as React from 'react';

export function getReportExportComponentForAssessment(props: CommandBarProps): JSX.Element {
    const {
        deps,
        assessmentStoreData,
        assessmentsProvider,
        featureFlagStoreData,
        tabStoreData,
    } = props;
    const reportGenerator = deps.reportGenerator;
    const reportExportComponentProps: ReportExportComponentProps = {
        deps: deps,
        exportResultsType: 'Assessment',
        reportGenerator: reportGenerator,
        pageTitle: tabStoreData.title,
        scanDate: deps.getCurrentDate(),
        htmlGenerator: description =>
            reportGenerator.generateAssessmentReport(
                assessmentStoreData,
                assessmentsProvider,
                featureFlagStoreData,
                tabStoreData,
                description,
            ),
        updatePersistedDescription: value =>
            props.deps.detailsViewActionMessageCreator.addResultDescription(value),
        getExportDescription: () => props.assessmentStoreData.resultDescription,
        featureFlagStoreData: props.featureFlagStoreData,
    };

    return <ReportExportComponent {...reportExportComponentProps} />;
}

export function getReportExportComponentForFastPass(props: CommandBarProps): JSX.Element {
    const scanResult = props.visualizationScanResultData.issues.scanResult;

    if (!scanResult) {
        return null;
    }

    const selectedTest = props.visualizationStoreData.selectedFastPassDetailsView;

    if (selectedTest !== VisualizationType.Issues) {
        return null;
    }

    const { deps, tabStoreData } = props;
    const scanDate = deps.getDateFromTimestamp(scanResult.timestamp);
    const reportGenerator = deps.reportGenerator;

    const reportExportComponentProps: ReportExportComponentProps = {
        deps: deps,
        scanDate: scanDate,
        reportGenerator: reportGenerator,
        pageTitle: tabStoreData.title,
        exportResultsType: 'AutomatedChecks',
        htmlGenerator: description =>
            reportGenerator.generateFastPassAutomatedChecksReport(
                scanDate,
                tabStoreData.title,
                tabStoreData.url,
                props.cardsViewData,
                description,
            ),
        updatePersistedDescription: () => null,
        getExportDescription: () => '',
        featureFlagStoreData: props.featureFlagStoreData,
    };

    return <ReportExportComponent {...reportExportComponentProps} />;
}
