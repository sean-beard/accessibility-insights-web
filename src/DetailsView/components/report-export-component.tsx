// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { escape } from 'lodash';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';
import { ReportGenerator } from 'reports/report-generator';
import { ExportResultType } from '../../common/extension-telemetry-events';
import { ExportDialog, ExportDialogDeps } from './export-dialog';

export type ReportExportComponentDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
} & ExportDialogDeps;

export interface ReportExportComponentProps {
    deps: ReportExportComponentDeps;
    exportResultsType: ExportResultType;
    reportGenerator: ReportGenerator;
    pageTitle: string;
    scanDate: Date;
    htmlGenerator: (descriptionPlaceholder: string) => string;
    updatePersistedDescription: (value: string) => void;
    getExportDescription: () => string;
    featureFlagStoreData: FeatureFlagStoreData;
}

export interface ReportExportComponentState {
    isOpen: boolean;
    exportName: string;
    exportDescription: string;
    exportData: string;
}

export class ReportExportComponent extends React.Component<
    ReportExportComponentProps,
    ReportExportComponentState
> {
    constructor(props) {
        super(props);
        this.state = {
            isOpen: false,
            exportName: '',
            exportDescription: '',
            exportData: '',
        };
    }

    private onDismissExportDialog = () => {
        this.setState({ isOpen: false });
    };

    private onExportDescriptionChange = (value: string) => {
        this.props.updatePersistedDescription(value);
        const escapedExportDescription = escape(value);
        this.setState({ exportDescription: escapedExportDescription });
    };

    private generateHtml = () => {
        const { htmlGenerator } = this.props;
        const exportData = htmlGenerator(this.state.exportDescription);
        this.setState({ exportDescription: '', exportData });
    };

    private onExportButtonClick = () => {
        const { reportGenerator, exportResultsType, scanDate, pageTitle } = this.props;
        const exportName = reportGenerator.generateName(exportResultsType, scanDate, pageTitle);
        const exportDescription = this.props.getExportDescription();
        this.setState({ exportDescription, exportName, isOpen: true });
    };

    public render(): JSX.Element {
        const { deps, exportResultsType } = this.props;
        const { isOpen, exportName, exportDescription, exportData } = this.state;
        return (
            <>
                <ActionButton iconProps={{ iconName: 'Export' }} onClick={this.onExportButtonClick}>
                    Export result
                </ActionButton>
                <ExportDialog
                    deps={deps}
                    isOpen={isOpen}
                    fileName={exportName}
                    description={exportDescription}
                    html={exportData}
                    onClose={this.onDismissExportDialog}
                    onDescriptionChange={this.onExportDescriptionChange}
                    exportResultsType={exportResultsType}
                    onExportClick={this.generateHtml}
                    featureFlagStoreData={this.props.featureFlagStoreData}
                />
            </>
        );
    }
}
