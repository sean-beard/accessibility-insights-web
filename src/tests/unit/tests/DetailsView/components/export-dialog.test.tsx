// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import { PrimaryButton } from 'office-ui-fabric-react';
import { Dialog } from 'office-ui-fabric-react';
import { TextField } from 'office-ui-fabric-react';
import * as React from 'react';
import { It, Mock, MockBehavior, Times } from 'typemoq';
import { FileURLProvider } from '../../../../../common/file-url-provider';
import { DetailsViewActionMessageCreator } from '../../../../../DetailsView/actions/details-view-action-message-creator';
import { ExportDialog, ExportDialogProps } from '../../../../../DetailsView/components/export-dialog';
import { FeatureFlags } from 'common/feature-flags';
import { FlaggedComponent } from 'common/components/flagged-component';

describe('ExportDialog', () => {
    const onCloseMock = Mock.ofInstance(() => {});
    const onDescriptionChangeMock = Mock.ofInstance((value: string) => {});
    const detailsViewActionMessageCreatorMock = Mock.ofType(DetailsViewActionMessageCreator, MockBehavior.Strict);
    const fileProviderMock = Mock.ofType<FileURLProvider>();
    const eventStub = 'event stub' as any;
    const onExportClickMock = Mock.ofInstance(() => {});
    let props: ExportDialogProps;

    beforeEach(() => {
        onCloseMock.reset();
        onDescriptionChangeMock.reset();
        detailsViewActionMessageCreatorMock.reset();
        onExportClickMock.reset();
        fileProviderMock.reset();

        const deps = {
            detailsViewActionMessageCreator: detailsViewActionMessageCreatorMock.object,
            fileURLProvider: fileProviderMock.object,
        };

        props = {
            deps,
            isOpen: false,
            html: 'fake html',
            fileName: 'THE REPORT FILE NAME',
            description: 'description',
            onClose: onCloseMock.object,
            onDescriptionChange: onDescriptionChangeMock.object,
            exportResultsType: 'Assessment',
            onExportClick: onExportClickMock.object,
            featureFlagStoreData: {}, // TODO change this
        };
    });

    describe('renders', () => {
        const isOpenOptions = [true, false];

        it.each(isOpenOptions)('with open %p', isOpen => {
            props.isOpen = isOpen;
            fileProviderMock
                .setup(provider => provider.provideURL(It.isAny(), It.isAnyString()))
                .returns(() => 'fake-url')
                .verifiable(Times.once());
            const wrapper = shallow(<ExportDialog {...props} />);
            expect(wrapper.getElement()).toMatchSnapshot();

            fileProviderMock.verifyAll();
        });
    });

    describe('user interaction', () => {
        it('closes the dialog onDismiss', () => {
            onCloseMock.setup(oc => oc()).verifiable(Times.once());
            fileProviderMock
                .setup(provider => provider.provideURL(It.isAny(), It.isAnyString()))
                .returns(() => 'fake-url')
                .verifiable(Times.once());
            onExportClickMock.setup(getter => getter()).verifiable(Times.never());
            const wrapper = shallow(<ExportDialog {...props} />);

            wrapper.find(Dialog).prop('onDismiss')();

            fileProviderMock.verifyAll();
            onCloseMock.verifyAll();
            onDescriptionChangeMock.verifyAll();
            detailsViewActionMessageCreatorMock.verifyAll();
            onExportClickMock.verifyAll();
        });

        it('handles click on export button', () => {
            const unchangedDescription = 'description';
            onDescriptionChangeMock.setup(dc => dc(It.isValue(unchangedDescription))).verifiable(Times.once());

            onCloseMock.setup(oc => oc()).verifiable(Times.once());
            fileProviderMock
                .setup(provider => provider.provideURL(It.isAny(), It.isAnyString()))
                .returns(() => 'fake-url')
                .verifiable(Times.exactly(2));
            onExportClickMock.setup(getter => getter()).verifiable(Times.once());

            detailsViewActionMessageCreatorMock
                .setup(a => a.exportResultsClicked(props.exportResultsType, props.html, eventStub))
                .verifiable(Times.once());

            const wrapper = shallow(<ExportDialog {...props} />);

            const flaggedComponent = wrapper.find(FlaggedComponent);

            flaggedComponent
                .dive()
                .find(PrimaryButton)
                .simulate('click', eventStub);

            fileProviderMock.verifyAll();
            onCloseMock.verifyAll();
            onDescriptionChangeMock.verifyAll();
            detailsViewActionMessageCreatorMock.verifyAll();
            onExportClickMock.verifyAll();
        });

        it('handles text changes for the description', () => {
            props.isOpen = true;
            fileProviderMock
                .setup(provider => provider.provideURL(It.isAny(), It.isAnyString()))
                .returns(() => 'fake-url')
                .verifiable(Times.once());
            const changedDescription = 'changed-description';
            onDescriptionChangeMock.setup(handler => handler(It.isValue(changedDescription))).verifiable(Times.once());

            const wrapper = shallow(<ExportDialog {...props} />);

            const textField = wrapper.find(TextField);
            textField.simulate('change', eventStub, changedDescription);

            fileProviderMock.verifyAll();
            onCloseMock.verifyAll();
            onDescriptionChangeMock.verifyAll();
            detailsViewActionMessageCreatorMock.verifyAll();
            onExportClickMock.verifyAll();
        });
    });
});
