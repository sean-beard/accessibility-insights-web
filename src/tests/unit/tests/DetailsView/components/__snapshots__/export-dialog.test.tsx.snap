// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportDialog renders with open false 1`] = `
<StyledWithResponsiveMode
  dialogContentProps={
    Object {
      "subText": "Optional: please describe the result (it will be saved in the report).",
      "title": "Provide result description",
      "type": 0,
    }
  }
  hidden={true}
  modalProps={
    Object {
      "containerClassName": "insights-dialog-main-override",
      "isBlocking": false,
    }
  }
  onDismiss={[Function]}
>
  <StyledTextFieldBase
    ariaLabel="Provide result description"
    autoFocus={true}
    multiline={true}
    onChange={[Function]}
    resizable={false}
    rows={8}
    value="description"
  />
  <StyledDialogFooterBase>
    <CustomizedPrimaryButton
      aria-roledescription="split button"
      download="THE REPORT FILE NAME"
      href="fake-url"
      menuProps={
        Object {
          "items": Array [
            Object {
              "key": "codepen",
              "onClick": [Function],
              "text": "CodePen",
            },
          ],
        }
      }
      onClick={[Function]}
      split={true}
      splitButtonAriaLabel="Export HTML to any of these format options"
      text="Export"
    />
  </StyledDialogFooterBase>
</StyledWithResponsiveMode>
`;

exports[`ExportDialog renders with open true 1`] = `
<StyledWithResponsiveMode
  dialogContentProps={
    Object {
      "subText": "Optional: please describe the result (it will be saved in the report).",
      "title": "Provide result description",
      "type": 0,
    }
  }
  hidden={false}
  modalProps={
    Object {
      "containerClassName": "insights-dialog-main-override",
      "isBlocking": false,
    }
  }
  onDismiss={[Function]}
>
  <StyledTextFieldBase
    ariaLabel="Provide result description"
    autoFocus={true}
    multiline={true}
    onChange={[Function]}
    resizable={false}
    rows={8}
    value="description"
  />
  <StyledDialogFooterBase>
    <CustomizedPrimaryButton
      aria-roledescription="split button"
      download="THE REPORT FILE NAME"
      href="fake-url"
      menuProps={
        Object {
          "items": Array [
            Object {
              "key": "codepen",
              "onClick": [Function],
              "text": "CodePen",
            },
          ],
        }
      }
      onClick={[Function]}
      split={true}
      splitButtonAriaLabel="Export HTML to any of these format options"
      text="Export"
    />
  </StyledDialogFooterBase>
</StyledWithResponsiveMode>
`;
